File: leetcode-go/Q051-Q100/Q086_Partition_List/Q086_Partition_List.go
package Q086_Partition_List

type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head

	gp, g := dummy, head
	for g != nil {
		for g != nil && g.Val < x {
			if g.Next == nil {
				return dummy.Next
			}
			g = g.Next
			gp = gp.Next
		}

		lp, l := gp, g
		for l != nil && l.Val >= x {
			if l.Next == nil {
				return dummy.Next
			}

			l = l.Next
			lp = lp.Next
		}

		lp.Next = l.Next
		gp.Next = l
		l.Next = g

		g = l
	}
	return dummy.Next
}

File: src/facebook/Q086_Partition_List.java
package facebook;

public class Q086_Partition_List {

    public ListNode partition(ListNode head, int x) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode curr = head;
        while (curr != null && curr.val < x) {
            curr = curr.next;
            prev = prev.next;
        }

        ListNode insert_prev = prev;
        while (curr != null) {
            if (curr.val < x) {
                ListNode next = curr.next;
                curr.next = insert_prev.next;
                insert_prev.next = curr;
                insert_prev = insert_prev.next;
                prev.next = next;
                curr = next;
            } else {
                prev = curr;
                curr = curr.next;
            }
        }
        return dummy.next;
    }

    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

}

